vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/MultiSelectListBox.java

```diff
@@ -66,8 +66,16 @@ public MultiSelectListBox() {
     private static <T> JsonArray modelToPresentation(
             MultiSelectListBox<T> listBox, Set<T> model) {
         JsonArray array = Json.createArray();
-        model.stream().map(listBox.getItems()::indexOf)
-                .forEach(index -> array.set(array.length(), index));
+        model.stream()
+                .filter(selectedItem -> listBox.getItems()
+                        .stream()
+                        .anyMatch(item ->
+                                listBox.getItemId(selectedItem)
+                                        .equals(listBox.getItemId(item))))
+                .forEach(itm -> {
```

Comment:
This `forEach` looks a bit extra. You already iterate through `listBox.getItems()` one line before. Could we filter these items from model which are in list box, and then add the record to json array immediately within the same iteration?

Reply:
AFAIK, no iteration happens reaching to intermediate operation such as `filter` until it reaches to some terminal operation like `forEach`. However, I will change the implementation since your point about `indexOf` depending on `eqauls` is correct and has to change.